extensions/src/main/scala/water/parser/CategoricalPreviewParseWriter.java

```diff
@@ -17,29 +17,30 @@
 
 package water.parser;
 
-import java.lang.reflect.Field;
-import water.util.IcedHashMap;
+public class CategoricalPreviewParseWriter {
 
-public class CategoricalPreviewParseWriter extends PreviewParseWriter {
+  public static byte guessType(String[] domain, int nLines, int nEmpty) {
+    final int nStrings = nLines - nEmpty;
+    final int nNums = 0;
+    final int nDates = 0;
+    final int nUUID = 0;
+    final int nZeros = 0;
 
-  public CategoricalPreviewParseWriter(String[] domain, int totalCount, int naCount) {
-    super(1);
-    this._nlines = totalCount;
-    this._nempty[0] = naCount;
-    this._nstrings[0] = totalCount - naCount;
-    IcedHashMap<String, String>[] domains = new IcedHashMap[1];
-    domains[0] = new IcedHashMapWrapper(domain);
-    setPrivateDomains(domains);
-  }
+    PreviewParseWriter.IDomain domainWrapper =
+        new PreviewParseWriter.IDomain() {
+          public int size() {
+            return domain.length;
+          }
+
+          public boolean contains(String value) {
+            for (String domainValue : domain) {
+              if (value.equals(domainValue)) return true;
```

Comment:
since this array can be pretty big what about doing sort and binary search after?

Reply:
@honzasterba I understand your motivation, but this method is called only when [the array size is 1](https://github.com/h2oai/h2o-3/pull/5086/files#diff-1f9b998f1c157b03f6fb652a49fa1460a79ec7a30e569563579f0d7326d027b4R181). I'm afraid if we sorted the array at the begging of `guessType` method, it would slow down most of the cases.

Reply:
For small amount of items, the slowdown introduced by sorting is absolutely negligible. For larger arrays, the search will affect performance heavily.

Reply:
agree with @mn-mikke - this is fine as-is

If we wanted to be uber-correct then we could replace 'contains' with 'isSingleValue'